design/service-binding.md

```diff
@@ -0,0 +1,356 @@
+spec:
+  kafka:
+    version: 2.4.0
+    replicas: 1
+    listeners:
+      plain: {}
+    storage:
+      type: ephemeral
+    zookeeper:
+      replicas: 1
+      storage:
+      type: ephemeral
+status:
+  listeners:
+  - type: plain
+    addresses:
+    - host: myhost.example.com
+      port: 9092
+```
+
+The Service Binding Operator looks for annotations starting `servicebinding`.
+
+This has the same shortcomings as the first example, with the advantage that the template code is not required in every single `ServiceBindingRequest`.
+
+It would be preferable if Strimzi custom resources were appropriately annotated without the user needing to do it.
+
+## Desired way of consuming binding information
+
+The aim is to make it easy to bind to any service using a `ServiceBindingRequest` CR. Here's an example:
+
+``` yaml
+apiVersion: service.binding/v1alpha1
+kind: ServiceBindingRequest
+metadata:
+  name: my-binding
+spec:
+  services:
+  - group: kafka.strimzi.io
+    kind: Kafka
+    version: v1beta1
+    resourceRef: my-cluster
+```
+
+Then the binding information is supplied in in a standard way, such as environment variables. Of course, in combination with developer tooling, this can remove the need for the developer to write the code to read credentials and so on. Instead, they're presented to the application as environment variables with well known names.
+
+## Binding data
+
+To connect to Strimzi, a service binding needs the followng:
+
+* **host** - hostname or IP address
+* **port** - port number
+* **userName** - username, optional
+* **password** - password or token, optional
+* **certificate** - certificate, optional
+
+These pieces of information can be provided in a variety of ways:
+
+* annotations on custom resources, indicating `spec` and `status` properties
+* ConfigMap
+* Secret
+
+## Strimzi enhancement - Concatenation of bootstrap server information
+
+The main difficulty with using the Service Binding Operator with Strimzi is that there's no convenient way to get a consolidated list of bootstrap servers for the listener that you wish to use.
+
+The way to find the bootstrap information from the `Kafka` CR is to look at the status for the listeners. Here's an example:
+
+``` yaml
+apiVersion:kafka.strimzi.io/v1beta1
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    version: 2.4.0
+    replicas: 2
+    listeners:
+      plain: {}
+    storage:
+      type: ephemeral
+  zookeeper:
+    replicas: 1
+    storage:
+      type: ephemeral
+status:
+  listeners:
+  - type: plain
+    addresses:
+    - host: myhost1.example.com
+      port: 9092
+    - host: myhost2.example.com
+      port: 9092
+```
+
+To bind nicely to this, ideally we'd have a consolidated list `myhost1.example.com:9092,myhost2.example.com:9092` that can directly be used as the Kafka client's bootstrap server list. Most examples dodge this issue by just using the first address in the list, but that's a terrible idea for availability.
```

Comment:
Again, I do not think this is true with Strimzi. The address we provide here is address to a service which load balances between all brokers. So you do not need to go through the complications to go through configuring multiple brokers. You just use the single service.

The only situation when we actually have multiple addresses for single listener is with the external listener of node port type where the whole situation is a bit more complicated. But I guess the service binding operator works only within the same cluster so this should not be relevant, right?

While adding the status, we were having a discussion about having a single field with `<host>:<port>` format or two separate fields. I guess adding a new field with both the hostname and port in one should be reasonable if it helps.

Reply:
Ah I guess this answers my previous question about why addresses is an array - it is due to the complications you mentioned around external listener and node ports. I think whether the service binding operator expects to be in the same Kubernetes cluster as the service it is binding to is a good question for us to ask the service binding team. I don't think we can safely make that assumption as it would be potentially useful if applications say on premise could bind with services in the cloud. I'll reach out to the service binding operator team and ask.

Reply:
If you bind the applications outside the cluster or in another cluster, you don't have anything what is discussed in this proposal available -> CRs, Secrets, etc. So I think the proposal would need to be very different.

Reply:
I think the key thing here is a consolidated string that can be used as bootstrap servers by a Kafka client. So, in the situations where Strimzi uses a list, build the list. In the situations where there's a single host:port pointing to a server, just concatenate them.

Reply:
The CR is an API. So I do not think randomly deciding whether to use one or the other is a good way. It needs to be easy to process and cosnume. So the question here is really whether we should:
* Add an addiitonal new form of showing the bootstrap servers
* Deprecate the list to have it removed in one of the future CRD versions

Reply:
I'd certainly favour the first of these. Wrt the deprecation, I'm not sure how aggressive Strimzi is about deprecation of these things.

Reply:
I personally do not have a problem with both of these. If we want to remove it in the future, we should ideally deprecate it first. I'm not sure whether anyone sees particular value in thos if we add the single string.

Reply:
From ServiceBinding operator team: "The operator assumes apps and services are in the same cluster. However, it’s technically possible to manually create a secret holding binding info for the external service from another cluster, then using SBO to bind the app to the secret (once #389  is implemented)" https://github.com/redhat-developer/service-binding-operator/issues/389

Reply:
So I think if we wanted to we could enable people to use the external listener, but probably not a priority

Reply:
Yes, but I'd prefer it to be listener-agnostic. Strimzi has plans to change how listeners are configured, so seems better to aim for flexibility now if we can.